File: lib/data/providers/scraping_provider.dart
import 'dart:async';
import 'package:dio/dio.dart';
import 'package:html/parser.dart' show parse;
import '../../domain/entities/mosque.dart';
import '../../domain/entities/prayer_times.dart';
import '../../domain/entities/geo_location.dart';
import 'prayer_data_provider.dart';

/// Provider C: HTML Scraping Fallback
/// Scrapes Mawaqit website for prayer times
class ScrapingProvider implements PrayerDataProvider {
  late Dio _dio;
  final _rateLimiter = _RateLimiter(minInterval: const Duration(seconds: 30));
  
  @override
  String get id => 'scraping';
  
  @override
  String get name => 'Web Scraping (Fallback)';
  
  @override
  String get description => 'Scrapes Mawaqit website directly (slower, use as fallback)';
  
  @override
  List<ConfigField> get configSchema => [
    const ConfigField(
      key: 'baseUrl',
      label: 'Mawaqit Base URL',
      type: ConfigFieldType.url,
      required: false,
      defaultValue: 'https://mawaqit.net',
      description: 'Base URL for Mawaqit website',
    ),
    const ConfigField(
      key: 'timeout',
      label: 'Request Timeout (seconds)',
      type: ConfigFieldType.number,
      defaultValue: '45',
    ),
    const ConfigField(
      key: 'enableCache',
      label: 'Enable HTML Caching',
      type: ConfigFieldType.boolean,
      defaultValue: 'true',
      description: 'Cache raw HTML for debugging',
    ),
  ];

  @override
  Future<void> initialize(Map<String, dynamic> config) async {
    final baseUrl = config['baseUrl'] as String? ?? 'https://mawaqit.net';
    final timeout = int.tryParse(config['timeout']?.toString() ?? '45') ?? 45;

    _dio = Dio(BaseOptions(
      baseUrl: baseUrl,
      connectTimeout: Duration(seconds: timeout),
      receiveTimeout: Duration(seconds: timeout),
      headers: {
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.9',
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
      },
    ));
  }

  @override
  Future<List<Mosque>> searchMosques(String query, {GeoLocation? location}) async {
    await _rateLimiter.throttle();
    
    try {
      // Mawaqit search URL
      final response = await _dio.get('/en/search', queryParameters: {
        'q': query,
        if (location != null) ...{
          'lat': location.latitude,
          'lng': location.longitude,
        },
      });

      if (response.statusCode == 200) {
        return _parseSearchResults(response.data as String);
      }
      
      throw ProviderException(
        message: 'Search failed: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<List<Mosque>> getNearbyMosques(GeoLocation location, {double radiusKm = 10}) async {
    await _rateLimiter.throttle();
    
    try {
      final response = await _dio.get('/en/nearby', queryParameters: {
        'lat': location.latitude,
        'lng': location.longitude,
        'radius': radiusKm,
      });

      if (response.statusCode == 200) {
        return _parseSearchResults(response.data as String);
      }
      
      throw ProviderException(
        message: 'Nearby search failed: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<PrayerTimes> getPrayerTimes(String mosqueId, {DateTime? date}) async {
    await _rateLimiter.throttle();
    
    try {
      final response = await _dio.get('/en/m/$mosqueId');

      if (response.statusCode == 200) {
        final html = response.data as String;
        return _parsePrayerTimes(html, mosqueId, date ?? DateTime.now());
      }
      
      throw ProviderException(
        message: 'Failed to fetch times: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<Mosque> getMosqueDetails(String mosqueId) async {
    await _rateLimiter.throttle();
    
    try {
      final response = await _dio.get('/en/m/$mosqueId');

      if (response.statusCode == 200) {
        return _parseMosqueDetails(response.data as String, mosqueId);
      }
      
      throw ProviderException(
        message: 'Failed to fetch details: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<bool> testConnection() async {
    try {
      final response = await _dio.get('/en');
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  List<Mosque> _parseSearchResults(String html) {
    final document = parse(html);
    final mosques = <Mosque>[];

    // Try multiple selectors for mosque listings
    final selectors = [
      '.mosque-item',
      '.mosque',
      '[data-mosque-id]',
      '.search-result',
      '.mosquee',
      '.card',
    ];

    for (final selector in selectors) {
      final elements = document.querySelectorAll(selector);
      if (elements.isNotEmpty) {
        for (final element in elements) {
          final id = element.attributes['data-id'] ??
              element.attributes['data-mosque-id'] ??
              _extractIdFromHref(element.querySelector('a')?.attributes['href']);
          
          final name = element.querySelector('.name, .title, h2, h3, h4')?.text.trim() ??
              element.text.split('\n').first.trim();
          
          final address = element.querySelector('.address, .location, .adresse')?.text.trim();
          
          if (id != null && name.isNotEmpty) {
            mosques.add(Mosque(
              id: id,
              name: name,
              address: address,
            ));
          }
        }
        break;
      }
    }

    return mosques;
  }

  String? _extractIdFromHref(String? href) {
    if (href == null) return null;
    final match = RegExp(r'/m/([^/]+)').firstMatch(href);
    return match?.group(1);
  }

  Mosque _parseMosqueDetails(String html, String mosqueId) {
    final document = parse(html);

    final name = document.querySelector('h1, .mosque-name, .page-title')?.text.trim() ??
        document.querySelector('[data-mosque-name]')?.attributes['data-mosque-name'] ??
        'Unknown Mosque';

    final address = document.querySelector('.address, .mosque-address, [data-address]')?.text.trim() ??
        document.querySelector('[data-address]')?.attributes['data-address'];

    final latStr = document.querySelector('[data-lat]')?.attributes['data-lat'] ??
        document.querySelector('[data-latitude]')?.attributes['data-latitude'];
    final lngStr = document.querySelector('[data-lng]')?.attributes['data-lng'] ??
        document.querySelector('[data-longitude]')?.attributes['data-longitude'];

    return Mosque(
      id: mosqueId,
      name: name,
      address: address,
      latitude: latStr != null ? double.tryParse(latStr) : null,
      longitude: lngStr != null ? double.tryParse(lngStr) : null,
    );
  }

  PrayerTimes _parsePrayerTimes(String html, String mosqueId, DateTime date) {
    final document = parse(html);

    // Look for prayer times in various formats
    final prayers = <String, Prayer>{};
    final prayerNames = ['Fajr', 'Dhuhr', 'Asr', 'Maghrib', 'Isha'];

    for (final name in prayerNames) {
      final prayer = _findPrayerTime(document, name, date);
      if (prayer != null) {
        prayers[name] = prayer;
      }
    }

    // Look for Jumuah
    final jumuah = _findPrayerTime(document, 'Jumuah', date) ??
        _findPrayerTime(document, 'Jumua', date) ??
        _findPrayerTime(document, 'Friday', date);

    if (prayers.length < 5) {
      throw ProviderException(
        message: 'Could not parse all prayer times from page',
        providerId: id,
      );
    }

    return PrayerTimes(
      date: date,
      mosqueId: mosqueId,
      cachedAt: DateTime.now(),
      fajr: prayers['Fajr']!,
      dhuhr: prayers['Dhuhr']!,
      asr: prayers['Asr']!,
      maghrib: prayers['Maghrib']!,
      isha: prayers['Isha']!,
      jumuah: jumuah,
    );
  }

  Prayer? _findPrayerTime(dynamic document, String prayerName, DateTime baseDate) {
    final selectors = [
      '[data-prayer="$prayerName"]',
      '[data-name="$prayerName"]',
      '.prayer-$prayerName',
      '#prayer-$prayerName',
      '.${prayerName.toLowerCase()}',
      '#${prayerName.toLowerCase()}',
    ];

    for (final selector in selectors) {
      final element = document.querySelector(selector);
      if (element != null) {
        final adhanTime = element.querySelector('.adhan, .athan, .time-start, [data-time="adhan"]')?.text.trim() ??
            element.attributes['data-adhan'] ??
            element.attributes['data-time'];

        final iqamaTime = element.querySelector('.iqama, .iqamah, .time-iqama, [data-time="iqama"]')?.text.trim() ??
            element.attributes['data-iqama'] ??
            element.attributes['data-iqamah'];

        if (adhanTime != null) {
          return Prayer(
            name: prayerName,
            adhan: _parseTimeString(adhanTime, baseDate),
            iqama: iqamaTime != null ? _parseTimeString(iqamaTime, baseDate) : null,
          );
        }
      }
    }

    return null;
  }

  DateTime _parseTimeString(String timeStr, DateTime baseDate) {
    final trimmed = timeStr.trim();
    final parts = trimmed.split(':');
    
    if (parts.length >= 2) {
      var hour = int.tryParse(parts[0]) ?? 0;
      final minute = int.tryParse(parts[1]) ?? 0;

      // Handle AM/PM
      if (trimmed.toLowerCase().contains('pm') && hour != 12) {
        hour += 12;
      } else if (trimmed.toLowerCase().contains('am') && hour == 12) {
        hour = 0;
      }

      return DateTime(baseDate.year, baseDate.month, baseDate.day, hour, minute);
    }

    return baseDate;
  }

  @override
  void dispose() {
    _dio.close();
  }
}

/// Simple rate limiter to avoid hammering servers
class _RateLimiter {
  final Duration minInterval;
  DateTime? _lastRequest;

  _RateLimiter({required this.minInterval});

  Future<void> throttle() async {
    if (_lastRequest != null) {
      final elapsed = DateTime.now().difference(_lastRequest!);
      if (elapsed < minInterval) {
        await Future.delayed(minInterval - elapsed);
      }
    }
    _lastRequest = DateTime.now();
  }
}

File: src-tauri/src/providers/scraping_provider.rs
use async_trait::async_trait;
use chrono::{NaiveDate, Utc};
use reqwest::Client;
use scraper::{Html, Selector};
use serde_json::Value;

use crate::models::*;
use crate::providers::{PrayerDataProvider, ProviderError, ProviderResult};

/// HTML Scraping Provider (Provider C)
/// Fallback that scrapes prayer times from mosque websites
pub struct ScrapingProvider {
    client: Client,
    base_url: Option<String>,
    rate_limit_delay_ms: u64,
}

impl ScrapingProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
                .build()
                .unwrap(),
            base_url: None,
            rate_limit_delay_ms: 1000, // 1 second between requests
        }
    }

    async fn rate_limit(&self) {
        tokio::time::sleep(tokio::time::Duration::from_millis(self.rate_limit_delay_ms)).await;
    }

    fn parse_time_from_text(&self, text: &str) -> Option<String> {
        // Try to find time patterns like "05:30", "5:30 AM", "17:30", etc.
        let re = regex::Regex::new(r"(\d{1,2}):(\d{2})\s*(AM|PM|am|pm)?").ok()?;
        re.captures(text).map(|cap| {
            let hour = cap.get(1)?.as_str();
            let minute = cap.get(2)?.as_str();
            Some(format!("{}:{}", hour, minute))
        })?
    }
}

#[async_trait]
impl PrayerDataProvider for ScrapingProvider {
    fn id(&self) -> &str {
        PROVIDER_SCRAPING
    }

    fn name(&self) -> &str {
        "HTML Scraping"
    }

    fn description(&self) -> &str {
        "Fallback provider that scrapes prayer times from mosque websites"
    }

    fn config_schema(&self) -> Vec<ConfigField> {
        vec![
            ConfigField::new("base_url", "Mosque Website URL", ConfigFieldType::Url)
                .description("The URL of the mosque's prayer times page"),
            ConfigField::new("rate_limit", "Rate Limit (ms)", ConfigFieldType::Number)
                .default_value("1000")
                .description("Delay between requests in milliseconds"),
        ]
    }

    async fn initialize(&mut self, config: Value) -> ProviderResult<()> {
        self.base_url = config["base_url"].as_str().map(|s| s.to_string());
        if let Some(delay) = config["rate_limit"].as_u64() {
            self.rate_limit_delay_ms = delay;
        }
        Ok(())
    }

    async fn search_mosques(&self, _query: &str, _location: Option<&GeoLocation>) -> ProviderResult<Vec<Mosque>> {
        // Scraping provider doesn't support searching
        Err(ProviderError::Other(
            "Scraping provider doesn't support mosque search".to_string(),
        ))
    }

    async fn get_nearby_mosques(&self, _location: &GeoLocation, _radius_km: f64) -> ProviderResult<Vec<Mosque>> {
        // Scraping provider doesn't support nearby search
        Err(ProviderError::Other(
            "Scraping provider doesn't support nearby search".to_string(),
        ))
    }

    async fn get_prayer_times(&self, _mosque_id: &str, date: Option<NaiveDate>) -> ProviderResult<PrayerTimes> {
        let base_url = self.base_url.as_ref().ok_or_else(|| {
            ProviderError::InvalidConfig("Base URL not configured".to_string())
        })?;

        self.rate_limit().await;

        let url = if let Some(d) = date {
            format!("{}?date={}", base_url, d.format("%Y-%m-%d"))
        } else {
            base_url.clone()
        };

        let response = self.client
            .get(&url)
            .send()
            .await
            .map_err(|e| ProviderError::Network(format!("Failed to fetch: {}", e)))?;

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Failed to fetch page".to_string(),
            });
        }

        let html = response.text().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to read response: {}", e))
        })?;

        let document = Html::parse_document(&html);

        // Try to find prayer times in the HTML
        // This is a generic implementation - specific selectors would depend on the site structure
        let prayer_names = ["Fajr", "Dhuhr", "Asr", "Maghrib", "Isha"];
        let mut prayers: Vec<(String, Option<String>)> = Vec::new();

        // Try common CSS selectors for prayer times
        let selectors_to_try = [
            "table.prayer-times td",
            ".prayer-time",
            "[data-prayer]",
            ".salah-time",
            ".adhan-time",
        ];

        for selector_str in &selectors_to_try {
            if let Ok(selector) = Selector::parse(selector_str) {
                let elements: Vec<_> = document.select(&selector).collect();
                if elements.len() >= 5 {
                    for (i, element) in elements.iter().enumerate().take(5) {
                        let text = element.text().collect::<String>();
                        if let Some(time) = self.parse_time_from_text(&text) {
                            prayers.push((prayer_names[i].to_string(), Some(time)));
                        }
                    }
                    break;
                }
            }
        }

        if prayers.len() < 5 {
            return Err(ProviderError::Parse(
                "Could not find prayer times in page".to_string(),
            ));
        }

        let date = date.unwrap_or_else(|| Utc::now().date_naive());
        let base_date = date.and_hms_opt(0, 0, 0).unwrap().and_local_timezone(Utc).unwrap();

        fn parse_time(date: chrono::DateTime<Utc>, time_str: &str) -> chrono::DateTime<Utc> {
            let parts: Vec<&str> = time_str.split(':').collect();
            let hour: u32 = parts.get(0).unwrap_or(&"0").parse().unwrap_or(0);
            let minute: u32 = parts.get(1).unwrap_or(&"0").parse().unwrap_or(0);
            date.date_naive().and_hms_opt(hour, minute, 0).unwrap().and_local_timezone(Utc).unwrap()
        }

        Ok(PrayerTimes {
            date: base_date,
            fajr: Prayer {
                name: "Fajr".to_string(),
                adhan: parse_time(base_date, prayers[0].1.as_deref().unwrap_or("05:00")),
                iqama: None,
                custom_rakah_count: None,
            },
            dhuhr: Prayer {
                name: "Dhuhr".to_string(),
                adhan: parse_time(base_date, prayers[1].1.as_deref().unwrap_or("12:00")),
                iqama: None,
                custom_rakah_count: None,
            },
            asr: Prayer {
                name: "Asr".to_string(),
                adhan: parse_time(base_date, prayers[2].1.as_deref().unwrap_or("15:00")),
                iqama: None,
                custom_rakah_count: None,
            },
            maghrib: Prayer {
                name: "Maghrib".to_string(),
                adhan: parse_time(base_date, prayers[3].1.as_deref().unwrap_or("18:00")),
                iqama: None,
                custom_rakah_count: None,
            },
            isha: Prayer {
                name: "Isha".to_string(),
                adhan: parse_time(base_date, prayers[4].1.as_deref().unwrap_or("19:30")),
                iqama: None,
                custom_rakah_count: None,
            },
            jumuah: None,
            mosque_id: Some("scraped".to_string()),
            mosque_name: None,
            cached_at: Some(Utc::now()),
        })
    }

    async fn test_connection(&self) -> ProviderResult<ProviderTestResult> {
        let start = std::time::Instant::now();

        match &self.base_url {
            Some(url) => {
                self.rate_limit().await;

                match self.client.get(url).send().await {
                    Ok(resp) => {
                        let latency = start.elapsed().as_millis() as u64;
                        if resp.status().is_success() {
                            Ok(ProviderTestResult {
                                success: true,
                                message: "Page accessible".to_string(),
                                latency_ms: Some(latency),
                            })
                        } else {
                            Ok(ProviderTestResult {
                                success: false,
                                message: format!("Page returned: {}", resp.status()),
                                latency_ms: Some(latency),
                            })
                        }
                    }
                    Err(e) => Ok(ProviderTestResult {
                        success: false,
                        message: format!("Connection failed: {}", e),
                        latency_ms: None,
                    }),
                }
            }
            None => Ok(ProviderTestResult {
                success: false,
                message: "No URL configured".to_string(),
                latency_ms: None,
            }),
        }
    }

    async fn get_mosque_details(&self, _mosque_id: &str) -> ProviderResult<Mosque> {
        Err(ProviderError::Other(
            "Scraping provider doesn't support mosque details".to_string(),
        ))
    }
}
